src/test/java/com/artipie/npm/proxy/NpmProxyITCase.java

```diff
@@ -0,0 +1,237 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2020 artipie
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.artipie.npm.proxy;
+
+import com.artipie.asto.Storage;
+import com.artipie.asto.memory.InMemoryStorage;
+import com.artipie.npm.proxy.http.NpmProxySlice;
+import com.artipie.vertx.VertxSliceServer;
+import com.google.common.net.MediaType;
+import io.vertx.reactivex.core.Vertx;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import org.apache.commons.configuration2.BaseConfiguration;
+import org.apache.commons.io.IOUtils;
+import org.hamcrest.MatcherAssert;
+import org.hamcrest.core.AllOf;
+import org.hamcrest.core.IsEqual;
+import org.hamcrest.core.StringContains;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockserver.client.MockServerClient;
+import org.mockserver.matchers.Times;
+import org.mockserver.model.BinaryBody;
+import org.mockserver.model.HttpRequest;
+import org.mockserver.model.HttpResponse;
+import org.testcontainers.Testcontainers;
+import org.testcontainers.containers.Container;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.MockServerContainer;
+
+/**
+ * Integration test for NPM Proxy.
+ *
+ * It uses MockServer container to emulate Remote registry responses,
+ * and Node container to run npm install command.
```

Comment:
npm is preinstalled on host machine, no need to run in a container

Reply:
It is pre-installed on the github workers, not on the developers machines

Reply:
Npm installation is a trivial task for a developer

Reply:
Why should we sacrifice maintainability and increase code complexity just for the sake of solving a task, which does not exist in real life?

Reply:
We increase the code complexity to make the environment reproducible. I think it's good tradeoff. We have several repository types, and I think that if will force the developers to install all required clients (npm, nuget, helm, rpm, docker itself, etc), that it will increase the complexity much more.

Reply:
npm client installation increases the complexity of the project onboarding just by one command and this is nothing, comparing to approach you to suggest: to create a container, which communicates with another container via an internal network.

The worst trade-off I have ever seen.

Reply:
No, you are mixing the points. The container with npm is used to avoid installation of this particular repository client. It is discussable in context of single project but we are supporting dozen of repositories with its own clients. This approach can be used to work with any of them. The alternative is to install every required client and skip some tests if the client is not available for your platform.
Containers communication is not related with it. Test requires some source of artifacts which can be accessed with HTTP. I use MockServer for this. It does not require container env and can be run directly from integration test. But:
1) without container we need to use mock server's own junit extentions. It can be difficult to support the test where 2 extensions are used in the same time.
2) the framework we already use allows to start the same MockServer. Why we need to bring second extension if we can do all we need with the one.